### PR TITLE
Fix #10229 EE10 Idle Timeout

### LINKED ISSUE #10229: HttpConfiguration.setIdleTimeout() breaks long running requests
**Jetty version(s)**
12 beta 4

**Java version/vendor** `(use: java -version)`
17

**OS type/version**
Windows

**Description**
I am not sure whether this might be intentional or not. But it looks more like a bug to me or at least like an unexpected change in Jetty 12.

From my understanding HttpConfiguration.idleTimeout refers to timeout in IO operations, not long running application code. In my code i was setting idleTimeoutout in Jetty 9 and 10 without any issues but with Jetty 12 long running request fail this error when the application code takes longer than the idleTimeout. Also note the suppressed NPE

<html><body>
<h2>HTTP ERROR 500 java.util.concurrent.TimeoutException: Idle timeout expired: 5015/5000 ms</h2>

URI: | http://localhost:8080/
-- | --
500
java.util.concurrent.TimeoutException: Idle timeout expired: 5015/5000 ms
java.util.concurrent.TimeoutException: Idle timeout expired: 5015/5000 ms

<h3>Caused by:</h3><pre>java.util.concurrent.TimeoutException: Idle timeout expired: 5015/5000 ms
	at org.eclipse.jetty.io.IdleTimeout.checkIdleTimeout(IdleTimeout.java:166)
	at org.eclipse.jetty.io.IdleTimeout.idleCheck(IdleTimeout.java:112)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
	Suppressed: java.lang.NullPointerException: Cannot invoke "java.util.function.Consumer.accept(Object)" because "onFailure" is null
		at org.eclipse.jetty.server.internal.HttpChannelState.lambda$onFailure$2(HttpChannelState.java:437)
		at org.eclipse.jetty.util.thread.SerializedInvoker$Link.run(SerializedInvoker.java:191)
		at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:969)
		at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1194)
		at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1149)
		... 1 more
</pre>
</body>
</html>

**How to reproduce?**

``` java
        public static void main(String... args) throws Exception
	{
		HttpServlet servlet = new HttpServlet()
		{
			@Override
			protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException
			{
				try
				{
					Thread.sleep(10000);
				}
				catch (InterruptedException e)
				{
					Thread.currentThread().interrupt();
				}

				resp.getWriter().printf("hello world");
			}
		};

		ServletContextHandler handler = new ServletContextHandler();
		handler.setContextPath("/");
		handler.addServlet(servlet, "/*");

		Server server = new Server();
		server.setHandler(handler);

		HttpConfiguration configuration = new HttpConfiguration();
		configuration.setIdleTimeout(5000);

		ServerConnector connector = new ServerConnector(server, new HttpConnectionFactory(configuration));
		connector.setPort(8080);
		server.setConnectors(new ServerConnector[]{connector});

		server.start();
	}
```

### PR DESCRIPTION
**Pull request by gregw**
Fix #10229

### CODE REVIEW

**Comment by gregw** [comment]
@sbordet I think the issue here is that we created the `Request.addIdleTimeoutListener` API, but then never got around to actually using it to ignore the idle timeouts as we need to whilst dispatched! Brain fade!!!!

**Comment by gregw** [comment]
Ah we do add a listener : `org.eclipse.jetty.ee10.servlet.ServletContextRequest#onIdleTimeout`
If I hack it to always return false, the test passes.  So something wrong in the impl.... I'll probably fix on my flight, so 10 hours until fix.

**Comment by gregw** [comment]
hmmmm ServletRequestState:
```java
    public boolean onIdleTimeout(TimeoutException timeout)
    {
        try (AutoLock ignored = lock())
        {
            if (LOG.isDebugEnabled())
                LOG.debug("onIdleTimeout {}", getStatusStringLocked(), timeout);
            // TODO this is almost always returning false?!? what about read/write timeouts???
            //      return _state == State.IDLE;
            return true;
        }
    }
```
yeah we didn't finish this!

If I implement the TODO, the test passes!

**Comment by gregw** [comment]
@sbordet @joakime I've pushed the fix suggested  by the TODO.  I'll look more during the flight, but it seams reasonable.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java [hunk 1]
```diff
@@ -353,23 +353,54 @@ public Runnable onIdleTimeout(TimeoutException t)
...
+                if (_failure == null)
+                {
+                    _failure = Content.Chunk.from(t);
+                }
+                else if (ExceptionUtil.areNotAssociated(_failure.getFailure(), t) && _failure.getFailure().getClass() != t.getClass())
+                {
+                    _failure.getFailure().addSuppressed(t);
+                }
+
+                return _serializedInvoker.offer(invokeOnContentAvailable, invokeWriteFailure);
```
Comment (gregw) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java]: Much of this was copied over from onFailure, so if there is an IO operation in progress, we just fail them when an idle timeout occurs.

I think this needs careful review

**Comment by gregw** [comment]
@lorban @sbordet  I've mode all the changes necessary to make idle timeouts for IO operations transient, but `AsyncContentProducer` has been written on the assumption that it can always re-read a failure.    So that class will need some reworking if we are ever to support transient idle timeouts.

**Review by sbordet — APPROVED** [review]

**Pull request merged**